src/main/java/com/google/api/generator/engine/ast/CastExpr.java

```diff
@@ -52,11 +52,15 @@ public CastExpr build() {
       // Only numeric types can be casted around.
       if (castType.isPrimitiveType() && exprType.isPrimitiveType()) {
         // Check that the types are either both booleans or both not booleans.
-        // TODO(xiaozhenliu): This is one of the places where we need to check the equality of boxed
-        // and primitive types.
         Preconditions.checkState(
             !(castType.equals(TypeNode.BOOLEAN) ^ exprType.equals(TypeNode.BOOLEAN)),
             "Numeric and boolean types are not inter-castable");
+      } else if (castType.isPrimitiveType() || exprType.isPrimitiveType()) {
+        // Check that if one type is primitive, and the other is its wrapper class type.
+        // For example: `Integer` and `int` they can be castable.
+        Preconditions.checkState(
+            TypeNode.isBoxedTypeEquals(castType, exprType),
```

Comment:
As per my comments in the doc, it may be best to alter `TypeNode.equals()` and use that instead of having an explicit `isBoxedTypeEquals` check here.

Reply:
sure, updated the `equals()` method and the condition check. Open to the error message throwing for the second check. Thanks!